commons-core/src/main/scala/com/avsystem/commons/serialization/macroCodecs.scala

```diff
@@ -101,12 +102,16 @@ abstract class ProductCodec[T <: Product](
   nullable: Boolean,
   fieldNames: Array[String]
 ) extends ApplyUnapplyCodec[T](typeRepr, nullable, fieldNames) {
-  final def writeObject(output: ObjectOutput, value: T): Unit = {
-    var i = 0
-    while (i < value.productArity) {
-      writeField(output, i, value.productElement(i))
-      i += 1
-    }
+  def size(value: T): Int = value.productArity
+
+  final def writeFields(output: ObjectOutput, value: T): Unit = {
+    val size = value.productArity
+    @tailrec def loop(idx: Int): Unit =
```

Comment:
You could just use `productIterator`

Reply:
I can avoid instantiating the iterator. You may consider this a premature optimization but I believe it to be no less readable than `productIterator` and it's _potentially_ faster than it.